test/Dialect/FIRRTL/layers-basic.mlir

```diff
@@ -0,0 +1,180 @@
+    }
+  }
+
+  firrtl.module @UncoloredDefineThroughSubfieldUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    // When looking at the layer requirements of a ref, we should look through
+    // bundle subfield ops.
+    %w = firrtl.wire : !firrtl.bundle<f: uint<1>>
+    firrtl.layerblock @A {
+      %f = firrtl.subfield %w[f] : !firrtl.bundle<f: uint<1>>
+      %r = firrtl.ref.send %f : !firrtl.uint<1>
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+
+  firrtl.module @CastAwayColorUnderLayerBlock(out %o : !firrtl.probe<uint<1>>) {
+    %w = firrtl.wire : !firrtl.probe<uint<1>, @A>
+  
+    // define %w, a wire colored by @A
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+    }
+  
+    // erase color from %w, safe under layerblock @A
+    firrtl.layerblock @A {
+      %r = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+      %d = firrtl.ref.resolve %r : !firrtl.probe<uint<1>>
+    }
+  }
+
+  firrtl.module @CastAwayPortColorUnderLayerBlock() {
+    firrtl.layerblock @A {
+      %o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>, @A>)
+      // safe to cast away @A from %o under layerblock @A
+      %r = firrtl.ref.cast %o : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+    }
+  }
+
+  firrtl.module @CastAwayColorUnderEnabledLayer(out %o : !firrtl.probe<uint<1>>) attributes {layers = [@A]} {
+    %w = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+      %d = firrtl.ref.resolve %s : !firrtl.probe<uint<1>>
+    }
+    // cast away the @A from %w, safe under enabled layers.
+    %r = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+    firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Enabled Layers Tests
+  //===--------------------------------------------------------------------===//
+
+  firrtl.module @CastAwayPortColorUnderEnabledLayer(out %o : !firrtl.probe<uint<1>>) attributes {layers = [@A]} {
+    %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>, @A>)
+    // safe to cast away @A from %foo_o under enabled layers
+    %r = firrtl.ref.cast %foo_o : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+    firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+  }
+
+  firrtl.module @DefineUncoloredProbeUnderEnabledLayerWithReferentUnderLayerBlock(out %o : !firrtl.probe<uint<1>>)
+  attributes {layers = [@A]} {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
```

Comment:
I'm glad this works and should be allowed. It would be good to, in a follow-up, add a pass that can inline this layerblock.